src/query/query_opfunc.c

```diff
@@ -401,29 +402,23 @@ qdata_copy_db_value_to_tuple_value (DB_VALUE * dbval_p, char *tuple_val_p, int *
 	  return ER_FAILED;
 	}
 
-      if ((DB_VALUE_DOMAIN_TYPE (dbval_p) == DB_TYPE_STRING || DB_VALUE_DOMAIN_TYPE (dbval_p) == DB_TYPE_VARNCHAR)
-	  && DB_GET_STRING_SIZE (dbval_p) >= PRIM_MINIMUM_STRING_LENGTH_FOR_COMPRESSION)
-	{
-	  /* Get max val_size from string */
-	  val_size = PRIM_STRING_MAXIMUM_DISK_SIZE (DB_GET_STRING_SIZE (dbval_p));
-
-	  OR_BUF_INIT (buf, val_p, val_size);
+      val_size = pr_data_writeval_disk_size (dbval_p);
+      OR_BUF_INIT (buf, val_p, val_size);
+      rc = (*(pr_type->data_writeval)) (&buf, dbval_p);
 
-	  rc = pr_get_size_and_write_string_to_buffer (&buf, val_p, dbval_p, &val_size, INT_ALIGNMENT);
-	}
-      else
+      if (rc != NO_ERROR)
 	{
-	  val_size = pr_data_writeval_disk_size (dbval_p);
-
-	  OR_BUF_INIT (buf, val_p, val_size);
-
-	  rc = (*(pr_type->data_writeval)) (&buf, dbval_p);
+	  /* ER_TF_BUFFER_OVERFLOW means that val_size or packing is bad. */
+	  assert (rc != ER_TF_BUFFER_OVERFLOW);
+	  return ER_FAILED;
 	}
 
+      /* Good moment to clear the compressed_string that might have been stored in the DB_VALUE */
+      rc = pr_clear_compressed_string (dbval_p);
```

Comment:
To minimize function calls:
```c
  if (dbval_type == DB_TYPE_VARCHAR || dbval_type == DB_TYPE_VARNCHAR)
    {
      rc = pr_clear_compressed_string (dbval_p);
      ...
    }
```

You may add a condition to check `dbval_p` requires to clear for more optimization.